rholang/src/main/scala/coop/rchain/rholang/interpreter/match.scala

```diff
@@ -0,0 +1,273 @@
+                      }
+                    })
+    } yield forcedYield
+
+  def spatialMatch(target: Par, pattern: Par): OptionalFreeMap[Unit] =
+    if (pattern.freeCount == 0) {
+      if (pattern == target)
+        StateT.pure(Unit)
+      else {
+        StateT.liftF(None)
+      }
+    } else {
+      /*
+        val varCount: Int = pattern.exprs.count({
+          case (EVar(FreeVar(_))) => true
+          case _                  => false
+        })
+        val wildcardCount: Int = pattern.exprs.count({
+          case (EVar(Wildcard)) => true
+          case _                => false
+        })
+       */
+      val varLevel: Option[Int] = possiblyFind[Expr, Int]({
+        case EVar(FreeVar(level)) => Some(level)
+        case _                    => None
+      }, pattern.exprs)
+      val wildcard: Boolean = exists[Expr]({
+        case EVar(Wildcard) => true
+        case _              => false
+      }, pattern.exprs)
+
+      def NoFrees(exprs: List[Expr]): List[Expr] =
+        exprs.filter({
+          case EVar(FreeVar(_)) => false
+          case EVar(Wildcard)   => false
+          case _                => true
+        })
+
+      for {
+        _ <- listMatchSingle[Send](target.sends,
+                                   pattern.sends,
+                                   spatialMatch,
+                                   (p, s) => p.prepend(s),
+                                   varLevel,
+                                   wildcard)
+        _ <- listMatchSingle[Expr](target.exprs,
+                                   NoFrees(pattern.exprs),
+                                   spatialMatch,
+                                   (p, e) => p.prepend(e),
+                                   varLevel,
+                                   wildcard)
+      } yield Unit
+    }
+
+  def spatialMatch(target: Channel, pattern: Channel): OptionalFreeMap[Unit] =
+    (target, pattern) match {
+      case (_, ChanVar(Wildcard)) => StateT.pure(Unit)
+      case (Quote(p), ChanVar(FreeVar(level))) => {
+        if (p.locallyFree.isEmpty)
+          StateT.modify(m => m + (level -> p))
+        else
+          StateT.liftF(None)
+      }
+      case (ChanVar(BoundVar(tlevel)), ChanVar(BoundVar(plevel))) =>
+        if (tlevel == plevel)
+          StateT.pure(Unit)
+        else
+          StateT.liftF(None)
+      case (Quote(tproc), Quote(pproc)) => spatialMatch(tproc, pproc)
+      case _                            => StateT.liftF(None)
+    }
+
+  def spatialMatch(target: Send, pattern: Send): OptionalFreeMap[Unit] =
+    if (target.persistent != pattern.persistent)
+      StateT.liftF(None)
+    else
+      for {
+        _           <- spatialMatch(target.chan, pattern.chan)
+        forcedYield <- foldMatch(target.data, pattern.data)
+      } yield forcedYield
+
+  def spatialMatch(target: Expr, pattern: Expr): OptionalFreeMap[Unit] =
+    (target, pattern) match {
+      case (t @ EList(tlist, _, _), p @ EList(plist, _, _)) => {
```

Comment:
`t` and `p` are not being used here, if compiler is not complaining, we should add wart here to block stuff like that

Reply:
Oh, they were being used to print while I was debugging.

Reply:
Done.